tsdb/record/record.go

```diff
@@ -117,6 +163,48 @@ func (d *Decoder) Series(rec []byte, series []RefSeries) ([]RefSeries, error) {
 	return series, nil
 }
 
+// Metadata appends metadata in rec to the given slice.
+func (d *Decoder) Metadata(rec []byte, metadata []RefMetadata) ([]RefMetadata, error) {
+	dec := encoding.Decbuf{B: rec}
+
+	if Type(dec.Byte()) != Metadata {
+		return nil, errors.New("invalid record type")
+	}
+	for len(dec.B) > 0 && dec.Err() == nil {
+		ref := dec.Uvarint64()
+		size := dec.Uvarint()
+
+		remainingBeforeReadFields := dec.Len()
+
+		typ := dec.Byte()
+		unit := dec.UvarintStr()
+		help := dec.UvarintStr()
+
+		// The bytes consumed will have shrunk, therefore delta between
+		// bytes unread before and bytes unread after is how much we consumed.
+		remainingAfterReadFields := dec.Len()
+		sizeFieldsRead := remainingBeforeReadFields - remainingAfterReadFields
+		if sizeFieldsUnread := size - sizeFieldsRead; sizeFieldsUnread > 0 {
+			// Need to skip fields that we didn't read and don't know about.
+			dec.Skip(sizeFieldsUnread)
+		}
```

Comment:
Could you please explain why do we need this logic or the `size` field? I can confidently say that we don't need either, but I might be missing something here.

Reply:
This is relevant to the question about the `size` field in the document describing the WAL format.

The logic around this field is about allowing new metadata types in the future, without having to strongly version the type of WAL entry that's being written, or needing a new parser if the Metadata struct changes. Here, `size` encodes how many metadata fields we know about so that when reading them back, we'll be able to skip over any fields we have no knowledge of.

This was based on [a discussion](https://github.com/prometheus/prometheus/pull/7771#discussion_r479443707) in the older PR, and no qualms were raised there so I followed the same road. I have no strong opinions on whether we should do that or simplify things a bit..

Actually, your and the team's input here would be valuable; do we realistically think that new Metadata types might be added in the future? If so how would we want to handle it in the WAL to ensure compatibility?